# PipeOne-NM

**PipeOne-NM** is a tool for genome functional annotation, non-coding RNA identification, transcripts alternative splicing analysis and gene differential expression in non-model organisms. 

PipeOne-NM is applicable to species with reference genome and can provide comprehensive insights into transcriptome of non-model organisms. It integrates twenty-one different tools to extract information from RNA-seq data and utilizes five R packages to carry out downstream analysis and data visualization. 

<p align="center">
<img src="https://github.com/Lisijun-m/pipeone-nm/blob/main/Figures/workflow.png" width="800px">
</p>

# Index
- [Prerequisites](#prerequisites)
- [Installation](#installation)
  - [1.Download PipeOne-NM](#1-download-pipeone-nm)
  - [2.Setup](#2-setup)
  - [3.Configuration](#3-configuration)
- [Run the Pipeline](#run-the-pipeline)
- [Results](#results) 
- [Rstudio scripts](#rstudio-scripts) 
- [Contact](#contact) 


## Prerequisites

1. [Docker](https://www.docker.com/) or [conda](https://docs.conda.io/en/latest/miniconda.html)
2. Java (version >= 1.7)
3. [Nextflow](https://www.nextflow.io/) (version >= 20.07.1.5413)

## Installation

### __1. Download PipeOne-NM__

```
https://github.com/nongbaoting/pipeone-nm.git
```

### __2. Setup__

##### 2.1 pull docker image or install conda environments

**Option 1: pull docker images run by dock**

```bash
docker push nongbaoting/pipeone_nm:latest
```

**Option 2: install conda environments** 

```
bash ./INSTALL/install.sh
```

##### __2.2 softwares need to be registeration__

* Trinotate
* tmhmm v2 (free academic download) [http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm](http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm)

* RNAMMER (free academic download) [http://www.cbs.dtu.dk/cgi-bin/sw_request?rnammer](http://www.cbs.dtu.dk/cgi-bin/sw_request?rnammer)
* signalP v4 (free academic download) [http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?signalp](http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?signalp)


### __3. Configuration__

Modify the program configuration file `conf/genomes.config`,  change the line below:

```
params {
  genomes {
  "fish_test" {
	refgenome = "/path/to/genome/genomic.fna"
	ref_miRNA = '/path/to/genome/public5/l/FishmiRNA-June2021-dre-mature.fasta'			  // for miranda
	uniprot = "/path/to/genome/uniprot_sprot.pep"                                                     // for Trinotate
	pfam = "/path/to/genome/Pfam-A.hmm"                                                               // for Trinotate
	sqlite = "/path/to/genome/Trinotate_20210616.sqlite"                                         	  // for Trinotate 
	min_length = 150
    }
  }
}
```

reqiured files which need for Trinotate, you could find this in [Trinotate ](https://github.com/Trinotate/Trinotate.github.io/blob/master/index.asciidoc#SequencesRequired)installatio instruction 

* refgenome
* ref_miRNA
* uniprot
* pfam
* sqlite

## __Run the Pipeline__


```
nextflow run /path/to/pipeOne-nm  -profile docker --sra './sra/*.sra' --genome fish_test
```

* Download RNA-seq data in sra format for test from NCBI
* Required Parameters
  * RNA-seq data
    * --sra, RNA-seq data in sra format
    * or --fastq, RNA-seq data in fastq format
  * --genome, genome version defined in `conf/genomes.config`,


## Results
The results generated by operating **PipeOne-NM** will be organized like the following.
Use grass carp RNA-seq data GSE185170 as analysis example.

* directory *sra*
  * sra format file downloaded from GEO. 
    * e.g. *SRR16151823.sra*
* directory *fasterq-dump*
  * fastq format file that has been converted by fastq-dump from sratoolkit. For pair end sequencing, each SRR corresponds to two files. 
    * e.g. *SRR16151823_1.fastq* and *SRR16151823_2.fastq*
* directory *fastp*
  * Quality control result file 
    * e.g. *SRR16151823_fastp.html* 
    * e.g. *SRR16151823_fastp.json* 
    * e.g. *SRR16151823_fastp.log*
  * fq.gz format file
    * e.g. *SRR16151823_1.fq.gz* and *SRR16151823_2.fq.gz*
* directory *hisat2*
  * directory *hisat2_index*
    * index file for genome alignment with HISAT2
      * e.g. *genomic.1.ht2*
  * bam format alignment file by hisat2
    * e.g. *SRR16151823.bam*
  * directory *un_conz_fastq*
    * e.g. *SRR16151823_um_1.fq.gz* and *SRR16151823_um_2.fq.gz*
* directory *stringtie*
  * gtf format file generated by stringtie
    * e.g. *SRR16151823.gtf*
* directory *TACO*
  * For more information about taco results, visit [https://tacorna.github.io/](https://tacorna.github.io/).
  * *assembly.gtf*
  * *assembly.bed*
  * *args.pickle*
  * *change_points.gtf*
  * *expr.neg.bedgraph*
  * *expr.none.bedgraph*
  * *expr.pos.bedgraph*
  * *path_graph_stats.txt*
  * *loci.txt*
  * *samples.txt*
  * *splice_graph.gtf*
  * *splice_junctions.bed*
  * *status.json*
  * *transfrags.bed*
  * *transfrags.filtered.bed*
  * *TACO.log*
* directory *gffread*
  * *Transcripts.fasta*
  * *Transctipts.fasta.gene_trans_map*
  * *GTFs.txt*
* directory *salmon*
  * For more information about salmon results, visit [https://combine-lab.github.io/salmon/](https://combine-lab.github.io/salmon/).
  * directory *salmon_index*
  * *matrix.log*
  * directory *samples*
    * *align_and_estimate_abundance.log*
    * quantify results for each sample (separately direcotry)
      * e.g. SRR16151823
    * *Transcripts.gene.counts.matrix*
    * *Transcripts.gene.TMM.EXPR.matrix*
    * *Transcripts.gene.TPM.not_cross_norm*
    * *Transcripts.gene.TPM.not_cross_norm.runTMM.R*
    * *Transcripts.gene.TPM.not_cross_norm.TMM_info.txt*
    * *Transcripts.isoform.counts.matrix*
    * *Transcripts.isoform.TMM.EXPR.matrix*
    * *Transcripts.isoform.TPM.not_cross_norm*
    * *Transcripts.isoform.TPM.not_cross_norm.runTMM.R*
    * *Transcripts.isoform.TPM.not_cross_norm.TMM_info.txt*
* directory *transdecoder*
  * For more information about transdecoder, visit [https://github.com/TransDecoder/TransDecoder/wiki](https://github.com/TransDecoder/TransDecoder/wiki).
  * *blastp.outfmt6*
  * *pfam.domtblout*
  * *Transcripts.fasta.transdecoder.bed*
  * *Transcripts.fasta.transdecoder.cds*
  * directory *Transcripts.fasta.transdecoder_dir*
  * directory *Transcripts.fasta.transdecoder_dir.__checkpoints*
  * directory *Transcripts.fasta.transdecoder_dir.__checkpoints_longorfs*
  * *TransDecoder.LongOrfs.log*
  * *TransDecoder.Predict.log*
* directory *trinotate*
  * For more information about trinotate results, visit [https://rnabio.org/module-07-trinotate/0007/02/01/Trinotate/](https://rnabio.org/module-07-trinotate/0007/02/01/Trinotate/).
  * *blastp.outfmt6*
  * *blastx.outfmt6*
  * *go_annotations.txt*
  * *signalp.out*
  * *tmhmm.out*
  * *Transcripts.fasta.rnammer.gff*
  * *transcriptsSuperScaffold.bed*
  * *transcriptsSuperScaffold.fasta*
  * *Trinotate_lncRNA.fa*
  * *Trinotate_lncRNA.fa_n50_stat*
  * *Trinotate_lncRNA.txt*
  * *Trinotate_mRNA.fa*
  * *Trinotate_mRNA.fa_n50_stat*
  * *Trinotate_mRNA.txt*
  * *TrinotatePFAM.out*
  * *Trinotate_rRNA.fa*
  * *Trinotate_rRNA.fa_n50_stat*
  * *Trinotate_rRNA.txt*
  * *Trinotate.xls*
* directory *bwaindex*
* directory *bwamem*
  * sam format alignment file by bwa-mem
    * e.g. *SRR16151823.sam*
* directory *ciri*
  * For more detailed description, please visit [https://ciri-cookbook.readthedocs.io/en/latest/CIRI2.html](https://ciri-cookbook.readthedocs.io/en/latest/CIRI2.html).
  * circRNA detection result based on sam format alignment file
    * e.g. *SRR16151823.ciri*
  * circRNA alternative splicing events detection
    * e.g. *SRR16151823.as_AS.list*
    * e.g. *SRR16151823.as_coverage.list*
    * e.g. *SRR16151823.as_jav.list*
    * e.g. *SRR16151823.as_library_length.list*
    * e.g. *SRR16151823.as.list*
    * e.g. *SRR16151823.as_splice.list*
  * Visualization of circRNA alternative splicing detection in pdf format
    * e.g. directory *SRR16151823*
  * directory *ciriquant*
    * directory *diff*
    * directory of each SRR, containing quantification result for each sample
* directory *bedtools*
  * fasta format circRNA extracted according to circRNA detection results
    * e.g. *SRR16151823.fasta*
* directory *miranda*
  * For more details about miRanda result, visit [https://cbio.mskcc.org/miRNA2003/miranda.html](https://cbio.mskcc.org/miRNA2003/miranda.html) for more information.
  * circRNA-miRNA interaction prediction file
    * e.g. *SRR16151823_circ_miRNA.txt*
* directory *rmats*
  * For details about rmats result, visit RMATS website [http://rnaseq-mats.sourceforge.net/](http://rnaseq-mats.sourceforge.net/) for more information.
  * A3SS.MATS.JCEC.txt*
  * *A3SS.MATS.JC.txt*
  * *A5SS.MATS.JCEC.txt*
  * *A5SS.MATS.JC.txt*
  * *fromGTF.A3SS.txt*
  * *fromGTF.A5SS.txt*
  * *fromGTF.MXE.txt*
  * *fromGTF.novelEvents.A3SS.txt*
  * *fromGTF.novelEvents.A5SS.txt*
  * *fromGTF.novelEvents.MXE.txt*
  * *fromGTF.novelEvents.RI.txt*
  * *fromGTF.RI.txt*
  * *fromGTF.SE.txt*
  * *JCEC.raw.input.A3SS.txt*
  * *JCEC.raw.input.A5SS.txt*
  * *JCEC.raw.input.MXE.txt*
  * *JCEC.raw.input.RI.txt*
  * *JCEC.raw.input.SE.txt*
  * *JC.raw.input.A3SS.txt*
  * *JC.raw.input.A5SS.txt*
  * *JC.raw.input.MXE.txt*
  * *JC.raw.input.RI.txt*
  * *JC.raw.input.SE.txt*
  * *MXE.MATS.JCEC.txt*
  * *MXE.MATS.JC.txt*
  * *RI.MATS.JCEC.txt*
  * *RI.MATS.JC.txt*
  * *SE.MATS.JCEC.txt*
  * *SE.MATS.JC.txt*

## Rstudio scripts
Compared with above process, differential expression analysis needs to be tailored to different datasets. It may involve removing of batch effect, differential analysis between multiple biological conditions and retrieving differential expression genes according to individualized threshold. 

Therefore, the differential analysis pipeline is not included in our PiprOne-NM, but here we present an example of performing differential analysis on grass carp circRNA host genes in folder Rstudio-example-scripts.

## Contact
For any problem, write to lisj35@mail2.sysu.edu.cn

